File: adapters/runtime/common/dom-adapter/audioContext/GainNode.js
import AudioNode from "./AudioNode";
import AudioParam from "./AudioParam";

class GainNode extends AudioNode {
    /**
     * @param context {BaseAudioContext} 音频模块
     * @param options 可选的 可用属性如下：
     *  gain 默认1，可选范围（-∞,+∞）
     */
    constructor(context, options) {
        super(context);

        this._gain = options && options.gain || new AudioParam();
    }

    get gain() {
        return this._gain;
    }
}

export default GainNode;

File: native/cocos/audio/graph_based/GainNode.cpp
#include "audio/graph_based/GainNode.h"
#include "LabSound/extended/AudioContextLock.h"
namespace cc {
GainNode::GainNode(BaseAudioContext* ctx, const GainNodeOptions& options) : AudioNode(ctx) {
    _node = std::make_shared<lab::GainNode>(*ctx->getInnerContext());
    if (options.gain) {
        _gain = std::shared_ptr<AudioParam>(AudioParam::createParam(std::dynamic_pointer_cast<lab::GainNode>(_node)->gain().get()));
    };
    lab::ContextGraphLock lck(_ctx->getInnerContext(), "initNode");
    if (options.channelCount.is_initialized()) {
        _node->setChannelCount(lck, options.channelCount.get()); 
    }
    _node->setChannelCountMode(lck, options.channelCountMode.value_or(lab::ChannelCountMode::Max));
    if (options.channelInterpretation.is_initialized()) {
        _node->setChannelInterpretation(options.channelInterpretation.get());
    }
}
AudioParam* GainNode::gain() {
    return _gain.get();
}

}
